example/src/main/scala/example/Example.scala

```diff
@@ -0,0 +1,66 @@
+package example
+
+import cats.Monad
+import cats.syntax.flatMap._
+import slog.{LoggerFactory, LoggingContext}
+
+class Example[F[_]](loggerFactory: LoggerFactory[F],
+                    loggingContext: LoggingContext[F])(implicit F: Monad[F]) {
+  private[this] val logger = loggerFactory.make[Example[F]]
+  // or
+  // private [this] val logger = loggerFactory.make("my-logger")
+
+  // example output json does not include timestamps, log levels, thread name,
+  // logger names etc to keep things simple. They are of course included in the real
+  // output.
+  def foo: F[Unit] = {
+    // simple log message. Might take from scope.
+    // {"message": "Hello world!!", "file": "Example.scala", "line": 19}
+    logger.info.msg("Hello world!!") >>
+      // Provide extract context to the log message
+      // {"message": "Hello", "file": "Example.scala", "line": 22, "correlation_id": "<VALUE>"}
+      logger.info.context("correlation_id", "<VALUE>").msg("Hello") >>
```

Comment:
calling `context` should go _before_ calling `info`
(the current order suggests there are different levels of context (`info`, `warn`,...), but that's not the case)

Reply:
> I looked at it and I like it! 🙂 Few things:
> 
>     * Do we really need the concept of `LoggerFactory`? Couldn't we just include the `implicit classTag` everywhere? Or make it similar to Metrics with something like `Logger.named` method which allows you to name your `Logger`? I know having the factory is common but to me it seems like unnecessary concept and what I don't like about it is that you pass the factory and you create a "dependency" (the logger) in the class itself instead of passing the logger directly to constructor. Just an idea, what do you think?

Well we don't need it, yet I didn't want to get too far from how logging is done right now. I'm actually not against it as long as there is big enough consensus on this.

>     * The macro for logging should do the log level check `if (isDebug)` but I guess that's just not finished yet.

It turns out it's not that easy if we want to be really pure. Current API tries to defer as much as possible, but there is still some allocation. And I'm not sure if macros will help in this case. (probably not)

>     * Do I understand it correctly that using `LoggingContext` and `logger.context` is essentially the same thing but `LoggingContext` allows the scoping? Again, do we need the `LoggingContext` and do we need to pass it as extra parameter to the class? It seems to me that this will be very cumbersome because you'll have like 2 or 3 constructor parameters yourself and then you'll have two for logging, one for monitoring and who knows what... I'd try to simplify this.

They look similar and even work similarly, but they are two different things. `LoggingContext` is used for context that is not really tied to a single log message. Typical example would be correlation id. On the other hand, `Logger.context` is used for log message specific context. That would typically be some case class.

It might seem that `Logger.context` is not really needed (which in fact is not), but I included it to simplify it for the user. And most importantly, it will allow us to provide more efficient implementation of this use case.

>     * Is there specific reason for the syntax `log.info.msg`? It's a bit unusual to me, I'm more used to `log.info(msg)` or something like that.

Not really, except that I found it more consistent with the rest.